File: VB/SimpleHtmlEdit/SimpleHtmlEdit.vb
Imports DevExpress.Utils.Text
Imports DevExpress.Utils.Text.Internal
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.ViewInfo
Imports System
Imports System.ComponentModel
Imports System.Diagnostics
Imports System.Drawing
Imports System.Windows.Forms

Namespace SimpleHtmlEditExample
	<UserRepositoryItem("RegisterSimpleHtmlEdit")>
	Public Class RepositoryItemSimpleHtmlEdit
		Inherits RepositoryItemTextEdit

		Public Sub New()
		End Sub
		Shared Sub New()
			RegisterSimpleHtmlEdit()
		End Sub

		Public Const CustomEditName As String = "SimpleHtmlEdit"
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return CustomEditName
			End Get
		End Property
        Public Shared Sub RegisterSimpleHtmlEdit()
            Dim img As EditImageIndexes = 0
            Dim eci As EditorClassInfo = New EditorClassInfo(CustomEditName, GetType(SimpleHtmlEdit), GetType(RepositoryItemSimpleHtmlEdit), GetType(SimpleHtmlEditViewInfo), New TextEditPainter(), True, img)
            EditorRegistrationInfo.Default.Editors.Add(eci)
        End Sub
    End Class

	<ToolboxItem(True)>
	Public Class SimpleHtmlEdit
		Inherits BaseEdit

		Protected Property PrevCursor() As Cursor
		Public Sub New()
		End Sub

		Protected Overrides Sub OnMouseMove(ByVal e As MouseEventArgs)
			MyBase.OnMouseMove(e)
			TryUpdateCursorByHyperlink(e)
		End Sub

		Protected Overridable Sub TryUpdateCursorByHyperlink(ByVal e As MouseEventArgs)
			If ViewInfo.GetStringInfo() Is Nothing Then
				Return
			End If
			Dim linkFound As StringBlock = ViewInfo.GetStringInfo().GetLinkByPoint(e.Location)
			If linkFound IsNot Nothing Then
				If PrevCursor Is Nothing Then
					PrevCursor = Cursor
				End If
				Cursor = Cursors.Hand
			Else
				If PrevCursor IsNot Nothing Then
					Cursor = PrevCursor
				End If
				PrevCursor = Nothing
			End If
		End Sub

		Protected Overrides Sub OnMouseDown(ByVal e As MouseEventArgs)
			Try
				Dim linkFound As StringBlock = ViewInfo.GetStringInfo().GetLinkByPoint(e.Location)
				If linkFound IsNot Nothing Then
					OpenUrlInDefaultBrowser(linkFound.Link)
				End If

			Catch exc As Exception
				MessageBox.Show("link is not valid")
			End Try
		End Sub

		Public Sub OpenUrlInDefaultBrowser(ByVal url As String)
			Process.Start(url)
		End Sub

		Shared Sub New()
			RepositoryItemSimpleHtmlEdit.RegisterSimpleHtmlEdit()
		End Sub
		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return RepositoryItemSimpleHtmlEdit.CustomEditName
			End Get
		End Property
		Protected Friend Shadows ReadOnly Property ViewInfo() As SimpleHtmlEditViewInfo
			Get
				Return TryCast(MyBase.ViewInfo, SimpleHtmlEditViewInfo)
			End Get
		End Property
	End Class

	Public Class SimpleHtmlEditViewInfo
		Inherits TextEditViewInfo

		Public Sub New(ByVal item As RepositoryItem)
			MyBase.New(item)
		End Sub

		Public Function GetStringInfo() As StringInfo
			Return StringInfo
		End Function

		Public Overrides Function GetMouseCursor(ByVal point As Point) As Cursor
			If StringInfo.GetLinkByPoint(point) IsNot Nothing Then
				Return Cursors.Hand
			End If

			Return MyBase.GetMouseCursor(point)
		End Function

		Public Overrides ReadOnly Property AllowHtmlString() As Boolean
			Get
				Return True
			End Get
		End Property
		Public Overrides ReadOnly Property IsRequiredUpdateOnMouseMove() As Boolean
			Get
				Return True
			End Get
		End Property
	End Class
End Namespace

File: CS/SimpleHtmlEdit/SimpleHtmlEdit.cs
using DevExpress.Utils.Text;
using DevExpress.Utils.Text.Internal;
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.ViewInfo;
using System;
using System.ComponentModel;
using System.Diagnostics;
using System.Drawing;
using System.Windows.Forms;
 
namespace SimpleHtmlEditExample {
    [UserRepositoryItem("RegisterSimpleHtmlEdit")]
    public class RepositoryItemSimpleHtmlEdit : RepositoryItemTextEdit {
        public RepositoryItemSimpleHtmlEdit() { }
        static RepositoryItemSimpleHtmlEdit() { RegisterSimpleHtmlEdit(); }
 
        public const string CustomEditName = "SimpleHtmlEdit";
        public override string EditorTypeName { get { return CustomEditName; } }
        public static void RegisterSimpleHtmlEdit() {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(CustomEditName,
              typeof(SimpleHtmlEdit), typeof(RepositoryItemSimpleHtmlEdit),
              typeof(SimpleHtmlEditViewInfo), new TextEditPainter(), true, null));
        }
    }
 
    [ToolboxItem(true)]
    public class SimpleHtmlEdit : BaseEdit {
        protected Cursor PrevCursor { get; set; }
        public SimpleHtmlEdit() { }
 
        protected override void OnMouseMove(MouseEventArgs e) {
            base.OnMouseMove(e);
            TryUpdateCursorByHyperlink(e);
        }
 
        protected virtual void TryUpdateCursorByHyperlink(MouseEventArgs e) {
            if(ViewInfo.GetStringInfo() == null) return;
            StringBlock linkFound = ViewInfo.GetStringInfo().GetLinkByPoint(e.Location);
            if(linkFound != null) {
                if(PrevCursor == null)
                    PrevCursor = Cursor;
                Cursor = Cursors.Hand;                
            } else {
                if(PrevCursor != null)
                    Cursor = PrevCursor;
                PrevCursor = null;
            }
        }

        protected override void OnMouseDown(MouseEventArgs e) {
            try {
                StringBlock linkFound = ViewInfo.GetStringInfo().GetLinkByPoint(e.Location);
                if(linkFound != null)
                    OpenUrlInDefaultBrowser(linkFound.Link);
 
            } catch(Exception exc) {
                MessageBox.Show("link is not valid");
            }
        }
 
        public void OpenUrlInDefaultBrowser(string url) {
            Process.Start(url);
        }
 
        static SimpleHtmlEdit() { RepositoryItemSimpleHtmlEdit.RegisterSimpleHtmlEdit(); }
        public override string EditorTypeName { get { return RepositoryItemSimpleHtmlEdit.CustomEditName; } }
        protected internal new SimpleHtmlEditViewInfo ViewInfo { get { return base.ViewInfo as SimpleHtmlEditViewInfo; } }
    }
 
    public class SimpleHtmlEditViewInfo : TextEditViewInfo {
        public SimpleHtmlEditViewInfo(RepositoryItem item) : base(item) { }
         
        public StringInfo GetStringInfo() {
            return StringInfo;
        }
 
        public override Cursor GetMouseCursor(Point point) {
            if(StringInfo.GetLinkByPoint(point) != null)
                return Cursors.Hand;
             
            return base.GetMouseCursor(point);
        }
 
        public override bool AllowHtmlString { get { return true; } }
        public override bool IsRequiredUpdateOnMouseMove { get { return true; } }
    }
}